Add connection to the Saline interface to MSTDB-TP

rks171 opened the pull request:
This adds a new fluid property class extension that wraps the Saline code, which is an interface to the MSTDB-TP salt fluid property databse.  Saline is added as an optional submodule.

closes #22724

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

bobsalko committed:
Add connection to the Saline interface to MSTDB-TP

This adds a new fluid property class extension that wraps the
Saline code, which is an interface to the MSTDB-TP salt fluid
property databse.  Saline is added as an optional submodule.

closes #22724

Update modules/fluid_properties/include/userobjects/SalineFluidProperties.h

Co-authored-by: Daniel Schwen <daniel@schwen.de>

Update modules/fluid_properties/Makefile

Co-authored-by: Daniel Schwen <daniel@schwen.de>

Update modules/fluid_properties/src/userobjects/SalineFluidProperties.C

Co-authored-by: Daniel Schwen <daniel@schwen.de>

Update modules/fluid_properties/src/userobjects/SalineFluidProperties.C

Co-authored-by: Daniel Schwen <daniel@schwen.de>

loganharbour commented on .gitmodules:
```diff
@@ -8,3 +8,6 @@
 [submodule "petsc"]
     path = petsc
     url = https://gitlab.com/petsc/petsc.git
+[submodule "modules/fluid_properties/saline_moose"]
+	path = modules/fluid_properties/saline_moose
+	url = git@code.ornl.gov:neams/saline_moose.git
```
```suggestion
[submodule "modules/fluid_properties/saline_moose"]
    path = modules/fluid_properties/saline_moose
    url = https://code.ornl.gov/neams/saline_moose.git
```

dschwen commented on .gitmodules:
```diff
@@ -8,3 +8,6 @@
     url = https://gitlab.com/petsc/petsc.git
+[submodule "modules/fluid_properties/saline_moose"]
+	path = modules/fluid_properties/saline_moose
+	url = git@code.ornl.gov:neams/saline_moose.git
```
No, let's just directly include the `saline` repo as a submodule and let's get rid of the `saline_moose` repo alltogether.

rks171 replied:
I am fine with that but will need guidance on how to get the `saline` repo working with the MOOSE build system.  That was my reason for repackaging it into `saline_moose` in the first place.

dschwen replied:
Yeah, you can look at my Thermocimica PR [here](https://github.com/idaholab/moose/pull/22712).

dschwen replied:
More specifically look at the [`thermochimica.mk`](https://github.com/idaholab/moose/pull/22712/files#diff-3c66f7d8738957757c65dc1e904ff253939e6b13a7f9834625b030a30f9c6f48) file (and where include it from [`chemical_reactions.mk`](modules/chemical_reactions/chemical_reactions.mk)).

On that note it would be cool if you could retain the same directory layout and have the saline submodule under `modules/fluid_properties/contrib/saline`.

dschwen replied:
with the build magic located in `modules/fluid_properties/contrib/saline.mk`

rks171 replied:
Should be fixed now

dschwen commented on modules/fluid_properties/Makefile:
```diff
@@ -18,9 +18,26 @@ ADDITIONAL_CPPFLAGS += -Wall -Wextra
 include $(FRAMEWORK_DIR)/build.mk
 include $(FRAMEWORK_DIR)/moose.mk
 
+################################## SALINE #######################################
+# Include link to Saline property database if the directory exists
+ifneq ("$(wildcard saline_moose/Makefile)", "")
+   #MOOSE_UNITY=false
+   APPLICATION_DIR    := $(shell pwd)/saline_moose
+   APPLICATION_NAME   := saline_moose
+   BUILD_EXEC         := no
+   ADDITIONAL_CPPFLAGS += -DSALINEAPP_H
+   ADDITIONAL_INCLUDES := -I$(APPLICATION_DIR)/include/base
+   ADDITIONAL_LIBS    += -L$(APPLICATION_DIR)/lib
+   DEP_APPS           := $(shell $(FRAMEWORK_DIR)/scripts/find_dep_apps.py $(APPLICATION_NAME))
+   include            $(FRAMEWORK_DIR)/app.mk
+   #MOOSE_UNITY=true
+endif
+
```
This won't work if you are not directly compiling the `fluid_properties` app. Put this stuff inside `fluid_properties.mk` instead.

rks171 replied:
Where is the `fluid_properties.mk` file located?

rks171 replied:
Nevermind... that's what I'm adding.

rks171 replied:
Fixed

dschwen commented on modules/fluid_properties/Makefile:
```diff
@@ -18,9 +18,26 @@ ADDITIONAL_CPPFLAGS += -Wall -Wextra
+   APPLICATION_DIR    := $(shell pwd)/saline_moose
+   APPLICATION_NAME   := saline_moose
+   BUILD_EXEC         := no
+   ADDITIONAL_CPPFLAGS += -DSALINEAPP_H
```
```suggestion
   ADDITIONAL_CPPFLAGS += -DSALINE_ENABLED
```

for consistency

dschwen commented on modules/fluid_properties/include/userobjects/SalineFluidProperties.h:
```diff
@@ -0,0 +1,109 @@
+
+#include "SinglePhaseFluidProperties.h"
+
+#ifdef SALINEAPP_H
```
```suggestion
#ifdef SALINE_ENABLED
```

dschwen commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,173 @@
+  return params;
+}
+
+// Input parser for salt component names
+std::vector<std::string>
+splitString_str(std::string str)
+{
+  std::string delimiter = "-";
+  std::string tmpstr = str;
+
+  std::vector<std::string> list;
+  size_t pos = 0;
+  std::string token;
+  while ((pos = tmpstr.find(delimiter)) != std::string::npos)
+  {
+    token = tmpstr.substr(0, pos);
+    list.push_back(token);
+    tmpstr.erase(0, pos + delimiter.length());
+  }
+  list.push_back(tmpstr);
+  return list;
+}
```
check out [`MooseUtils::tokenize`](https://github.com/idaholab/moose/blob/next/framework/include/utils/MooseUtils.h#L719-L746) . Better than rolling your own.

rks171 replied:
Switched over to `tokenize`

dschwen commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,173 @@
+}
+
+// Input parser for salt component mole fractions
+std::vector<Real>
+splitString_float(std::string str)
+{
+  std::vector<std::string> list = splitString_str(str);
+  std::vector<Real> res;
+  for (std::string s : list)
+    res.push_back(stof(s));
+  return res;
+}
```
same here. There is a tokenize version that does the type conversion `tokenizeAndConvert`

rks171 replied:
Just saw this, but this was also fixed.

GiudGiud replied:
has been addressed

dschwen commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,173 @@
+  : SinglePhaseFluidProperties(parameters)
+{
+
+#ifdef SALINEAPP_H
```
```suggestion
#ifdef SALINE_ENABLED
```

dschwen commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,173 @@
+  Real mole_sum = 0.0;
+  for (Real val : valList)
+    mole_sum = mole_sum + val;
+  if (abs(mole_sum - 1.0) > 1e-6)
```
```suggestion
  if (std::abs(mole_sum - 1.0) > 1e-6)
```

we always explicitly specify the namespace for the math functions

dschwen commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,173 @@
+#endif
+}
+
+// Unit conversion constants for communicating with Saline
```
we do have an entire unit conversion library (for future reference)

rks171 replied:
If you point me to it, I will use that instead.

dschwen replied:
Let's not worry about it for this PR. But it is at https://github.com/idaholab/moose/blob/next/framework/include/utils/Units.h with docs [here](https://mooseframework.inl.gov/source/utils/Units.html)

dschwen commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,173 @@
+
+#ifdef SALINEAPP_H
+
+std::string
```
you might want to return a `const std::string &` as best practice (we've been bitten by this before...)

dschwen replied:
Oh, that's an overnight method? Then nevermind.

dschwen commented on modules/fluid_properties/test/tests/saline/test.i:
```diff
@@ -0,0 +1,70 @@
+    value_a = ${fparse (2.579-6.24e-4*700)*1000}
+    value_b = "rho"
+    absolute_tolerance = "0.1" # kg/m^3
+  [../]
```
you can remove all `../` and `./`. We also have a VSCode plugin that does the input formatting/prettifying for you

rks171 replied:
Fixed

dschwen commented:
I'll be on leave for the next two weeks, but @logan or @GiudGiud should be able to help you. I would prefer it if we use the same submodule approach (layout and make files) as we do in the thermochimica PR.

bobsalko committed:
Updates for MR

bobsalko committed:
Run clang

Update modules/fluid_properties/fluid_properties.mk

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update modules/fluid_properties/include/userobjects/SalineFluidProperties.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

GiudGiud commented on modules/fluid_properties/include/userobjects/SalineFluidProperties.h:
```diff
@@ -0,0 +1,107 @@
+  /**
+   * Fluid name
+   *
+   * @return "flibe"
```
```suggestion
   * @return the name of the salt
```

GiudGiud reviewed (commented):
I m on travel now, I ll be more available next week

I ll try this with some of the TH codes in use

rks171 commented on modules/fluid_properties/test/tests/saline/saline_custom.prp:
```diff
@@ -0,0 +1,4 @@
+// System      , id , Mol Mass , Mol Frac             , Melt(K) , Var(%) , Reference   , Boil(K) , Var(%) , Reference      , rho_a     , rho_b     , range(K)      , Var(%) , Reference     , mu1_a     , mu2_b     , mu2_a      , mu2_b      , mu2_c     , range(K)      , Var(%) , Reference    , k_a        , k_b        , range(K)     , Var(%) , Reference   , cp_a      , cp_b       , cp_c       , cp_d  ...
+//
+//
```
You mean the file or the header in the file?  This is the database file that Saline reads in.

GiudGiud replied:
all good then

rks171 commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,168 @@
+std::string
+SalineFluidProperties::fluidName() const
+{
+  return _propName;
```
I was aiming to leave it more general in case there are ever non-salt fluids in Saline.  What about `_fluid_name`?

GiudGiud replied:
_fluid_name works

bobsalko committed:
Updates for MR

Update modules/fluid_properties/test/tests/saline/test.i

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update modules/fluid_properties/test/tests/saline/test.i

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Apply suggestions from code review

Co-authored-by: Daniel Schwen <daniel@schwen.de>

Apply suggestions from code review

rks171 commented on modules/fluid_properties/include/userobjects/SalineFluidProperties.h:
```diff
@@ -0,0 +1,107 @@
+
+protected:
+#ifdef SALINE_ENABLED
+  saline::Default_Data_Store _d;
```
Added

rks171 commented on modules/fluid_properties/contrib/saline.mk:
```diff
@@ -0,0 +1,37 @@
+# source files
+Saline_srcfiles    := $(shell find $(SALINE_SRC) -name "*.cc")
+#$(shell find $(SALINE_SRC) -name "*.cpp")
```
Removed

GiudGiud commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,168 @@
+  dcp_dp = 0;
+  dcp_dT = (cp_from_p_T(pressure, temperature + 1.0) - cp) / 1.0;
+}
+
+// Real
+// SalineFluidProperties::enthalpy_from_T(Real temperature) const
+//{
+//   return _tp.h_t_kg(temperature);
+// }
+//
+// Real
+// SalineFluidProperties::T_from_enthalpy(Real enthalpy) const
+//{
+//   return _tp.t_h_kg(enthalpy);
+// }
```
it would be h_from_p_T

rks171 replied:
Do you mean this is a function I should be overriding?  I do not see one named `h_from_p_T` in the `SinglePhaseFluidProperties` or `FluidProperties` classes.

GiudGiud replied:
It s present under a macro in SingleFluidProperties.h
 propfunc(h, p, T)
then for example
```
./FlinakFluidProperties.h:  virtual void h_from_p_T(Real p, Real T, Real & h, Real & dh_dp, Real & dh_dT) const override;
```

GiudGiud replied:
sounds good. T_from_p_h is a bit more niche btw, I think it s used by Pronghorn - subchannel

GiudGiud commented on modules/fluid_properties/test/tests/saline/test.i:
```diff
@@ -0,0 +1,70 @@
+[]
+
+[FluidProperties]
+  [flibe]
```
```suggestion
  [salt]
```

GiudGiud commented on modules/fluid_properties/test/tests/saline/test.i:
```diff
@@ -0,0 +1,70 @@
+    variable = rho
+    p = p
+    T = T
+    fp = flibe
```
```suggestion
    fp = salt
```

GiudGiud commented on modules/fluid_properties/fluid_properties.mk:
```diff
@@ -0,0 +1,7 @@
+# Saline
+SALINE_DIR ?= $(APPLICATION_DIR)/contrib/saline
```
this should be fluid properties module instead of APP_DIR ?

dschwen replied:
Hm. I can double check when I'm back. I thought that get changed multiple times during build to whatever it needs to be...

GiudGiud replied:
this should be changed to
```suggestion
SALINE_DIR ?= $(MOOSE_DIR)/modules/fluid_properties/contrib/saline
```

GiudGiud commented on modules/fluid_properties/fluid_properties.mk:
```diff
@@ -0,0 +1,7 @@
+SALINE_SRC = $(SALINE_DIR)/src
+ifneq ($(wildcard $(SALINE_SRC)/cpp/data_store.cc),)
+	ADDITIONAL_CPPFLAGS += -DSALINE_ENABLED
+	include $(APPLICATION_DIR)/contrib/saline.mk
```
same

dschwen replied:
No, see my comment below

rks171 replied:
The one with the three items?  I believe these are all addressed.

dschwen replied:
If those are addressed how come I can still see the use of `APPLICATION_DIR` her eon github?

dschwen replied:
you co
```suggestion
	include $(SALINE_DIR)/../saline.mk
```

GiudGiud commented on modules/fluid_properties/contrib/saline.mk:
```diff
@@ -0,0 +1,28 @@
+
+ADDITIONAL_LIBS += $(SALINE_DIR)/lib/libSaline-$(METHOD).la
+
+app_INCLUDES += -I$(SALINE_SRC) -I$(SALINE_SRC)/api -I$(SALINE_DIR)/include -I$(SALINE_SRC)/fortran
```
```suggestion
app_INCLUDES += -I$(SALINE_SRC) -I$(SALINE_DIR)/include
```
these folders do not exist?

rks171 replied:
I'm not sure where the `api` one came from.  But the fortran one does exist.  Does the build still work when you remove it?  There are header files in there, which is why I added it.

GiudGiud replied:
I might not have looked closely enough

rks171 replied:
I removed the `api` one since it is not there, but left the `fortran` one.

dschwen replied:
that api folder is probably just copy paste from the thermochimica makefile

GiudGiud commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,168 @@
+  return params;
+}
+
+//// Input parser for salt component names
+// std::vector<std::string>
+// splitString_str(std::string str)
+//{
+//   std::string delimiter = "-";
+//   std::string tmpstr = str;
+//
+//   std::vector<std::string> list;
+//   size_t pos = 0;
+//   std::string token;
+//   while ((pos = tmpstr.find(delimiter)) != std::string::npos)
+//   {
+//     token = tmpstr.substr(0, pos);
+//     list.push_back(token);
+//     tmpstr.erase(0, pos + delimiter.length());
+//   }
+//   list.push_back(tmpstr);
+//   return list;
+// }
+//
+//// Input parser for salt component mole fractions
+// std::vector<Real>
+// splitString_float(std::string str)
+//{
+//   std::vector<std::string> list = splitString_str(str);
+//   std::vector<Real> res;
+//   for (std::string s : list)
+//     res.push_back(stof(s));
+//   return res;
+// }
```
remove if unused

GiudGiud reviewed (changes requested):
It doesnt seem to compile from a module that depends on fluid_properties, such as NavierStokes or the combined module. See my comments in the makefile. 
```
libtool:   error: cannot find the library '/Users/giudgl/projects/moose_v4/modules/navier_stokes/contrib/saline/lib/libSaline-devel.la' or unhandled argument '/Users/giudgl/projects/moose_v4/modules/navier_stokes/contrib/saline/lib/libSaline-devel.la'
make: *** [/Users/giudgl/projects/moose_v4/framework/app.mk:416: /Users/giudgl/projects/moose_v4/modules/navier_stokes/navier_stokes-devel] Error 1
```
it misplaces the contrib into the local module instead of fluid_properties


Another question is do we want to silence compilation warnings from Saline?
@permcody @dschwen @loganharbour for opinions on silencing warnings from contrib

bobsalko committed:
Updates for merge request
refs #22724
Clang updates

Update modules/fluid_properties/fluid_properties.mk

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1254503/) on f239ee1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22725/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22725/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format af38e9f0b6976df8d0a90b1c6b339bef1dbc9b00`

loganharbour reviewed (dismissed):
I need to check CIVET recipes

GiudGiud commented:
@dschwen was looking at multi-level includes (including Saline in FP, and running another module that builds with FP)

dschwen commented:
> What else needs to be done to close this out?

- [ ]  `$(APPLICATION_DIR)/` -> `$(MOOSE_DIR)/modules/fluid_properties/`
- [ ]  fix clang formatting
- [ ] fix merge conflict in `.gitmodules`

let me know if you need any help with this

dschwen reviewed (dismissed):
> What else needs to be done to close this out?

- [x]  `$(APPLICATION_DIR)/` -> `$(MOOSE_DIR)/modules/fluid_properties/`
- [x]  fix clang formatting
- [x] fix merge conflict in `.gitmodules`

let me know if you need any help with this

GiudGiud commented:
Let's squash these commits together a bit, especially the "update some file" commits.

Where are we on the documentation failure with thermochimica? Do we still need all recipes to check it out? If so, is that going to be ok for saline too?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1297239/) on 83689c0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22725/site)

This comment will be updated on new commits.

GiudGiud commented:
The build is also failing because we are passing some strict flags from MOOSE (-Werror-unused etc) on Saline

dschwen commented:
> Where are we on the documentation failure with thermochimica? Do we still need all recipes to check it out? If so, is that going to be ok for saline too?

That is fixed.

dschwen commented on modules/fluid_properties/include/userobjects/SalineFluidProperties.h:
```diff
@@ -0,0 +1,139 @@
+class SalineFluidProperties : public SinglePhaseFluidProperties
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Woverloaded-virtual"
```
that makes me a bit uneasy

GiudGiud replied:
yeah this is not new, it s something we will fix everywhere when we get to fixing it

dschwen commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,146 @@
+//
+#include "SalineFluidProperties.h"
+
+registerMooseObject("FluidPropertiesApp", SalineFluidProperties);
```
Check out https://github.com/idaholab/moose/pull/23019/files for the approach taken for the optional Thermochimica coupling. There we also always register objects, but we indicate a missing Thermochimica library in the class description, too.

rks171 replied:
Are you talking about the calls to `checkLibraryAvailability` and associated error handling?

GiudGiud replied:
there's also a addClassDescription.

@dschwen from our discussion there, should we copy-paste this ThermochimicaUtils as a SalineUtils?
I think @rks171 achieved most of the same results with just preprocessor directives directly in the object. I think making a util made more sense for thermochimica because there was more than one object

GiudGiud replied:
I reproduced the other behavior that the utils provide in the object. The checkAvailability was already done

GiudGiud replied:
resolving

dschwen commented on modules/fluid_properties/src/userobjects/SalineFluidProperties.C:
```diff
@@ -0,0 +1,146 @@
+  Real mole_sum = 0.0;
+  for (Real val : valList)
+    mole_sum = mole_sum + val;
+  if (std::abs(mole_sum - 1.0) > 1e-6)
```
I'm not a fan of a hardcoded custom tolerance here. You could use `MooseUtils::absoluteFuzzyEqual(mole_sum, 1.0)`

rks171 replied:
This will be included in the next set of changes.

GiudGiud replied:
@rks171

GiudGiud replied:
fixed

dschwen commented:
> The build is also failing because we are passing some strict flags from MOOSE (-Werror-unused etc) on Saline

What a wonderful opportunity to fix the Saline code base :-D

GiudGiud commented:
> > The build is also failing because we are passing some strict flags from MOOSE (-Werror-unused etc) on Saline
> 
> What a wonderful opportunity to fix the Saline code base :-D

GiudGiud commented:
A few options here:
- Saline gets modified to remove the warning
- #pragma GCC diagnostic ignored "-Woverloaded-virtual" get added to Saline 
- we pass additional flags to undo the MOOSE ones

dschwen commented on modules/fluid_properties/test/tests/saline/tests:
```diff
@@ -0,0 +1,11 @@
+    rel_err = 1e-8
+    allow_test_objects = true
+    threading = '!pthreads'
+    required_objects = 'SalineFluidProperties'
```
This needs to be changed
```suggestion
    required_submodule = 'contrib/saline'
```

Because the object now always exists (albeit in a defunct state if the module is not available).

rks171 replied:
I will include this with the other changes.

dschwen commented:
> My preference would be to not request changes from the Saline team if that is a possible avenue.

That was not the intention of my comment. I might make that pr to saline myself.

HendersonSC commented:
> > My preference would be to not request changes from the Saline team if that is a possible avenue.
> 
> That was not the intention of my comment. I might make that pr to saline myself.

Many of the Saline's compiler warnings have been addressed. There is an unused parameter which is reserved for pressure. This parameter may become required at a later date. If compiler options equivalent GCC's combination of `-Werror -Wall -Wextra` then the equivalent `-Wno-unused-parameter` will be required to compile Saline directly.

GiudGiud committed:
Update saline submodule
refs #22724

GiudGiud commented:
That's great news
I just rebased and updated the submodule, see where we stand

GiudGiud commented:
We are still hitting the errors.

I just squashed unnecessary commits and tried adding another flag to build Saline. 
This may not be quite what we want, I ll have to check how far this flag propagates

dschwen commented:
Just comment out the name is the unused parameter.

GiudGiud commented:
something like :   function(other arguments..... , Real /*pressure*/) will do here

GiudGiud committed:
Restrict test to having the saline submodule
refs #22724

GiudGiud commented:
ok we cant do what I did [here](cc8074fa3c7693e2449563a8473909355c56e43a), it propagates to the whole repo

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1442159/) on 0688bad wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>40cceb</th>
      <th colspan=4>#22725 0688ba</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.43%</td>
      <td>85.36%</td>
      <td>-0.07%</td>
      <td>45.45%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6868</td>
      <td>6909</td>
      <td>+41</td>
      <td>25</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1171</td>
      <td>1185</td>
      <td>+14</td>
      <td>30</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22725/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22725/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22725/coverage/xfem)

</details>

## Warnings

- `fluid_properties` new line coverage rate 45.45% is less than the suggested 90.0%



This comment will be updated on new commits.

HendersonSC commented:
Yes that will suppress the warning, but I am not sure that Saline intends to suppress the warning at this time. Either way I believe that is a different philosophical discussion and the problem at hand is MOOSE's using the perfectly valid form of Saline which is already published. 

Make provides for target specific variables, so the following may work:
`saline : ADDITIONAL_CPPFLAGS += -DSALINE_ENABLED -Wno-unused-parameter`.

I am not really proficient in Make and am definitely not versed in MOOSE's flavoring of Make. So the above may need some massaging.

GiudGiud commented on modules/fluid_properties/fluid_properties.mk:
```diff
@@ -0,0 +1,7 @@
+ifneq ($(wildcard $(SALINE_SRC)/cpp/data_store.cc),)
+	ADDITIONAL_CPPFLAGS += -DSALINE_ENABLED -Wno-unused-parameter
+	include $(APPLICATION_DIR)/contrib/saline.mk
+endif
```
```suggestion
endif
saline: ADDITIONAL_CPPFLAGS += -Wno-unused-parameter
```

GiudGiud commented:
I agree let's figure this make command out to enable the flag very locally. 
In the meantime a few things to consider:
- line coverage is very low. The way we get near 100% coverage in fluid properties is with a unit test. It makes it very easy to call each routine and check their output. See anything in fluid_properties/unit/src
- the current set of functions implemented likely only suffices for SAM, NSFV will want more AD routines, subchannel will want more some p_rho variable set routines, THM will want some (v,e) routines. While we dont need to do this now, I m raising awareness so we can have a plan

GiudGiud commented:
There is no saline target though in the Makefile. I think all the code is compiled under the same target. 
@permcody any idea how to move forward here

GiudGiud committed:
Add unit testing to SalineFP, refs #22724

GiudGiud commented on modules/fluid_properties/Makefile:
```diff
@@ -24,3 +24,5 @@ include $(MODULE_DIR)/modules.mk
 
 ###############################################################################
 # Additional special case targets should be added here
+
+
```
```suggestion
```

GiudGiud commented:
Test suite is failing because documentation is missing. I ll catch the formatting things in another commit but someone more qualified should write the actual SalineFluidProperties documentation (in doc/content/source/fluidproperties)

GiudGiud committed:
Add stub for saline fluid properties
Provide SQA requirement and issues in saline regression test
Adapt to new directory structure

GiudGiud committed:
Rename to SalineMoltenSalt as Saline is too close to NaCl English name

GiudGiud committed:
Add documentation to Saline fluid properties

GiudGiud commented:
Testing already includes Saline.
@dschwen if you could please review, I have too many commits here

rks171 commented:
Is there anything additional I need to do to get this MR pushed in?

GiudGiud commented:
Needs a reviewer, then there may be additional comments to address.
@idaholab/moose-ccb

lindsayad commented on .gitmodules:
```diff
@@ -12,6 +12,9 @@
     update = none
+[submodule "modules/fluid_properties/contrib/saline"]
+	path = modules/fluid_properties/contrib/saline
+	url = https://code.ornl.gov/neams/saline
```
I don't know if we want this checked out by default?

rks171 replied:
probably safe not to.  Just set `update = none`?

lindsayad replied:
yea

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+SalineMoltenSaltFluidProperties::SalineMoltenSaltFluidProperties(const InputParameters & parameters)
+  : SinglePhaseFluidProperties(parameters)
+{
+
```
```suggestion
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+{
+
+#ifdef SALINE_ENABLED
+
```
```suggestion
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+
+#ifdef SALINE_ENABLED
+
+  std::string propDef = getParam<std::string>("prop_def");
```
```suggestion
  const auto & propDef = getParam<std::string>("prop_def");
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+  _d.load(propDef);
+  bool success = _tp.initialize(&_d);
+  if (!success)
+    mooseError("The initialize of the Saline interface has failed");
```
```suggestion
    mooseError("The initialization of the Saline interface has failed");
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+  bool success = _tp.initialize(&_d);
+  if (!success)
+    mooseError("The initialize of the Saline interface has failed");
+  std::string name = getParam<std::string>("comp_name");
+  std::string comp = getParam<std::string>("comp_val");
```
```suggestion
  const auto & name = getParam<std::string>("comp_name");
  const auto & comp = getParam<std::string>("comp_val");
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+  std::vector<Real> valList;
+  MooseUtils::tokenizeAndConvert<Real>(comp, valList, "-");
+  Real mole_sum = 0.0;
+  for (Real val : valList)
```
```suggestion
  for (const auto val : valList)
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+  MooseUtils::tokenizeAndConvert<Real>(comp, valList, "-");
+  Real mole_sum = 0.0;
+  for (Real val : valList)
+    mole_sum = mole_sum + val;
```
```suggestion
    mole_sum += val;
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+  if (!success)
+    mooseError("The composition set has failed");
+  _fluid_name = name;
+
+#endif
+#ifndef SALINE_ENABLED
+
```
```suggestion
#else
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+
+  mooseError("Saline was not made available during the build and can not be used. Make sure you "
+             "have the contrib/saline submodule checked out.");
+
```
```suggestion
```

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+}
+
+// Unit conversion constants for communicating with Saline
+const Real kPa_to_Pa = 1.0e3;
+const Real Pa_to_kPa = 1.0 / kPa_to_Pa;
+const Real kg_to_g = 1.0e3;
+const Real g_to_kg = 1.0 / kg_to_g;
+const Real m_to_cm = 1.0e2;
+const Real cm_to_m = 1.0 / m_to_cm;
+const Real N_to_mN = 1.0e3;
+const Real mN_to_N = 1 / N_to_mN;
```
I think these should either be `constexpr` or `static` class data members. At a minimum I would put these in an anonymous namespace but even that can run into collisions because of unity builds

rks171 replied:
You want to declare these as `constexpr` and also put them all into an anonymous namespace in this file?

lindsayad replied:
Or make them class data members

lindsayad commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,149 @@
+{
+  enthalpy = h_from_p_T(0.0, temperature);
+  dh_dp = 0.0;
+  // finite difference approximate
```
```suggestion
  // finite difference approximation
```

lindsayad commented on modules/fluid_properties/test/tests/saline/test.i:
```diff
@@ -0,0 +1,69 @@
+  type = GeneratedMesh
+  dim = 1
+  nx = 1
+
```
```suggestion
```

lindsayad commented on modules/fluid_properties/test/tests/saline/test.i:
```diff
@@ -0,0 +1,69 @@
+    variable = rho
+    nodeid = 0
+  []
+
```
```suggestion
```

bobsalko committed:
Address Alex's review
- add more constness and more autos
- dont always checkout saline
- move some constants to be constexpr

Update modules/fluid_properties/test/tests/saline/test.i

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

Update modules/fluid_properties/test/tests/saline/test.i

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1767071/) on 0c1ea40 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22725/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22725/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1acda23805be34b178665b209b2988bba734b183`

lindsayad commented:
@rks171 thank you for addressing the comments. I think this is just about ready to go as far as I'm concerned. Appreciate the contribution! Would you be able to squash your most recent commits into one review-addressing commit?

loganharbour commented:
I'll look at our checkouts in CIVET for this this evening.

GiudGiud commented:
I ll do the squash

lindsayad commented:
If it's a pain not having the new submodule get checked out when people run `git submodule update`, I'm supportive of making it non-optional. But I figure it would be good to be consistent with how we treat thermochimica. If we make this non-optional, then perhaps we should do so with thermochimica as well

GiudGiud commented:
optional is the way to go.
I ll make a PR to the recipes to include it

loganharbour commented:
> optional is the way to go.
> 
> I ll make a PR to the recipes to include it

Already doing it, as I said above.

loganharbour commented:
Recipes should be good

moosebuild commented:
All [jobs](https://civet.inl.gov/event/144225/) on edf427e : invalidated by @GiudGiud

Testing recipes

dschwen commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,133 @@
+{
+  InputParameters params = SinglePhaseFluidProperties::validParams();
+  params.addClassDescription("Molten salt fluid properties using Saline");
+  params.addRequiredParam<std::string>(
+      "comp_name", "The name of the components in the salt (e.g., 'LiF-NaF-KF')");
+  params.addRequiredParam<std::string>(
+      "comp_val", "The mole fraction of each salt component (e.g., '0.5-0.1-0.4)");
```
```suggestion
  params.addRequiredParam<std::string>(
      "comp_name", "The name of the components in the salt (e.g., 'LiF-NaF-KF')");
  params.addRequiredParam<std::string>(
      "comp_val", "The mole fraction of each salt component (e.g., '0.5-0.1-0.4)");
```

I'd much rather see those as type `std::vector<std::string>` and `std::vector<Real>`. I understand that this is the format saline expects, but it is unexpected for MOOSE users. You can easily `Mosoe::stringify` these vectors with a custom delimiter ('-`) to generate teh Saline compatible argument string.

dschwen replied:
I'd rather be the one to resolve

GiudGiud replied:
done. Please resolve!

GiudGiud replied:
resolving

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1776012/) on beb320a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22725/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22725/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 83eb82b88ae81d34bcd6e727749071fdb503f51b`

GiudGiud commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,133 @@
+  params.addClassDescription("Molten salt fluid properties using Saline");
+  params.addRequiredParam<std::vector<std::string>>("comp_name",
+                                                    "The name of the components in the salt");
+  params.addRequiredParam<std::vector<Real>`>("comp_val",
```
```suggestion
  params.addRequiredParam<std::vector<Real>>("comp_val",
```

GiudGiud commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,133 @@
+  bool success = _tp.initialize(&_d);
+  if (!success)
+    mooseError("The initialization of the Saline interface has failed");
+  const auto & name = getParam<std::string>("comp_name");
```
```suggestion
  const auto & name = getParam<std::vector<std::string>>("comp_name");
```

dschwen commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,133 @@
+  if (!success)
+    mooseError("The initialization of the Saline interface has failed");
+  const auto & name = getParam<std::string>("comp_name");
+  const auto & comp = getParam<std::string>("comp_val");
```
That's not going to work. You need to update the code here and below to the new parameter type. Instead of tokenize you will now have to stringify to build the `-` separated string for saline.

GiudGiud replied:
fixed

rks171 commented:
Just checking in on this MR again and hoping we can push this to completion.  Can someone go through and resolve any outstanding comments that are no longer issues so I can get a list of remaining issues to fix?

lindsayad commented:
@dschwen @GiudGiud please mark conversations as resolved (if they are resolved) if you haven't already done so

GiudGiud commented:
I think there's an outstanding item requested by Daniel and I started making some comments towards it but we never finished. 

We definitely should finish this PR, I was presenting at ANS on fluid properties in MOOSE and pitched our coming integration with Saline as our solution for molten salt properties :)

GiudGiud commented:
I can work on this next week if you don't get to it before that @rks171

GiudGiud committed:
Address Daniel s review:
- do not use a new format for specifying a vector

GiudGiud commented:
Let's see how this does. My only reservation is that I seem to be the only one to want this is in MOOSE. At least on this thread and in my recent interactions, though I do not make the MSR meeting every week. 

@tanoret @mtnwalker please confirm you will use this for your analysis, or @rks171 if you know of other customers. i know Saline already caters to many people using it in SAM, but they have integrated it there

lindsayad commented:
I would say that if a downstream application is using it and someone has contributed it upstream (here to MOOSE), then that is more than enough reason for me. If someone finds it useful, then others in the general community may too

rks171 commented:
@GiudGiud, the Mole team has also integrated Saline into their code using the same approach that I used for SAM, so at the least, this will prevent duplication in one MOOSE app.

GiudGiud commented:
@dschwen when you have time

GiudGiud commented:
@lindsayad @joshuahansel I need a volunteer to revive this.

joshuahansel commented:
@GiudGiud Ok, I assigned myself. @loganharbour are you happy with how we're doing the submodule here? I'm confident to review other aspects.

GiudGiud commented on modules/fluid_properties/src/fluidproperties/SalineMoltenSaltFluidProperties.C:
```diff
@@ -0,0 +1,132 @@
+  : SinglePhaseFluidProperties(parameters)
+{
+#ifdef SALINE_ENABLED
+  const auto & propDef = getParam<std::string>("prop_def");
```
```suggestion
  const auto & propDef = getParam<std::string>("prop_def_file");
```

GiudGiud commented on modules/fluid_properties/test/tests/saline/test.i:
```diff
@@ -0,0 +1,67 @@
+    type = SalineMoltenSaltFluidProperties
+    comp_name = "LiF NaF KF"
+    comp_val = "0.465 0.115 0.42"
+    prop_def = "saline_custom.prp"
```
```suggestion
    prop_def_file = "saline_custom.prp"
```

GiudGiud commented on modules/fluid_properties/unit/include/SalineMoltenSaltFluidPropertiesTest.h:
```diff
@@ -0,0 +1,35 @@
+    InputParameters uo_pars = _factory.getValidParams("SalineMoltenSaltFluidProperties");
+    uo_pars.set<std::vector<std::string>>("comp_name") = {"LiF", "NaF", "KF"};
+    uo_pars.set<std::vector<Real>>("comp_val") = {0.465, 0.115, 0.42};
+    uo_pars.set<std::string>("prop_def") = "../test/tests/saline/saline_custom.prp";
```
```suggestion
    uo_pars.set<std::string>("prop_def_file") = "../test/tests/saline/saline_custom.prp";
```

GiudGiud committed:
Address Josh's review
- better parameter name
- better documentation
Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

GiudGiud commented:
review should be addressed

joshuahansel commented on modules/fluid_properties/doc/content/source/fluidproperties/SalineMoltenSaltFluidProperties.md:
```diff
@@ -0,0 +1,42 @@
+!listing modules/fluid_properties/test/tests/saline/test.i block=FluidProperties
+
+It leverages the salt property input data in the `saline_custom.prp` file below.
+The data is sorten in columns. The first column is the name of the salt, then its id, then
```
```suggestion
The data is sorted in columns. The first column is the name of the salt, then its id, then
```

joshuahansel commented on modules/fluid_properties/doc/content/source/fluidproperties/SalineMoltenSaltFluidProperties.md:
```diff
@@ -0,0 +1,42 @@
+It leverages the salt property input data in the `saline_custom.prp` file below.
+The data is sorten in columns. The first column is the name of the salt, then its id, then
+a number of thermo-physical properties with additional metadata such as the range of validity,
+the uncertainty and a short name for the reference publication providing the data.
```
This explanation still does not give enough information for someone to make their own prp. The example is very useful, but there are a lot of columns that I still don't understand. Is there documentation on this file format somewhere else that we can link to? That would be fine; otherwise, I think we need to spell out the requirements for the format in greater detail, like a table with column number to the description of what goes there (with units). Also, is there always just a single row in this file? How do those reference strings get used?

GiudGiud replied:
all good questions. i think we ll want to reference somewhere else here

joshuahansel commented on modules/fluid_properties/doc/content/source/fluidproperties/SalineMoltenSaltFluidProperties.md:
````diff
@@ -0,0 +1,42 @@
+- viscosity
+- thermal diffusivity
+
+!alert note
+Saline must be installed alongside MOOSE using the `contrib/saline` submodule in the fluid properties
+module to be able to use this fluid property object. To initialize/download the saline submodule:
+
+```bash
+git submodule update --init contrib/saline
+```
````
The code block gets rendered outside of the alert block. To fix, change "!alert" to "!alert!" and then add a line with "!alert-end!" after your code block.

dschwen commented on modules/fluid_properties/contrib/saline.mk:
```diff
@@ -0,0 +1,28 @@
+# Include dependencies (see note above)
+-include $(Saline_deps)
+
+ADDITIONAL_LIBS += $(SALINE_DIR)/lib/libSaline-$(METHOD).la
```
This needs to be updated. Look at `thermochimica.mk`

GiudGiud replied:
done

GiudGiud committed:
Address Daniel's review
- fix makefile
- use different class description if not compiled with Saline
- use absoluteFuzzyEqual for checking molar sum

joshuahansel commented:
> review should be addressed

@GiudGiud Still have 4 unresolved comments by @dschwen (at least some did not appear to be resolved)

GiudGiud commented:
I resolved them and addressed the ones that were not addressed

GiudGiud commented:
447 days, this does not take the crown for longest running PR (that's probably general field transfers with 23 months :O)

GiudGiud commented:
Thanks for the review @joshuahansel
